### PR TITLE
New fuzzing platform

### LINKED ISSUE #1448: Integrate new fuzzing platform for Java
**Description**

We have new implementation of fuzzing using other concepts. We need to merge new implementation of Java for the plugin into main branch and test it before release. In case something goes wrong we should have an ability to revert this.

**Context**

New fuzzing platform elaborates implementation for new languages by reducing amount of the code to support new languages.

### PR DESCRIPTION
**Pull request by Markoutte**
# Description

Adds new fuzzing platform that simplifies developing fuzzing for new languages. Old implementation is still in codebase but will be deprecated and removed soon. Only base Java implementation migrated to the new fuzzing platform at the moment.

Please, see [Fuzzing Platform Design doc](https://github.com/UnitTestBot/UTBotJava/blob/pelevin/fuzzing_platform/docs/Fuzzing%20Platform.md) for understanding of basic concepts. In nutshell now any work with fuzzing requires some 'seeds' or 'tasks' which can be 4 types:

1. Simple seed is just a value with optional mutator
2. Known [structure] seed is something that has common structure for different types, e.g. integers (byte, short, int, long) which can be represented as a bit vector. This type can be used instead of 1 to faster developing of defaults such as integers, floats, strings, etc.
3. Recursive seed can represent some recursive structures that require to fuzz some parameters before constructing itself.
4. Collection seed represents collections such as arrays, lists, sets, maps and others.

There are [some demos](https://github.com/UnitTestBot/UTBotJava/tree/pelevin/fuzzing_platform/utbot-fuzzing/src/main/kotlin/org/utbot/fuzzing/demo) that clarify these concepts.

This platform is fully implemented for Java and plugin can be tested with [these samples](https://github.com/UnitTestBot/UTBotJava/tree/pelevin/fuzzing_platform/utbot-fuzzing/src/test/kotlin/org/utbot/fuzzing/samples).

Fixes #1448

Also note, that after short discussion this implementation doesn't use mocks because it looks wrong concepts in terms of fuzzing. Fuzzing doesn't analyze the code therefore it cannot supply correct mock calls for tests. Therefore #747 is discontinued starting this PR.

## Type of Change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

## Regression and integration tests

[Test samples](https://github.com/UnitTestBot/UTBotJava/tree/pelevin/fuzzing_platform/utbot-fuzzing/src/test/kotlin/org/utbot/fuzzing/samples) for manual testing were added. Integration test for fuzzing is being under developing.

## Automated Testing

All tests from the `utbot-fuzzer` and `utbot-fuzzing` modules must pass.

## Manual Scenario 

[Samples](https://github.com/UnitTestBot/UTBotJava/tree/pelevin/fuzzing_platform/utbot-fuzzing/src/test/kotlin/org/utbot/fuzzing/samples) for manual scenario.

# Checklist:

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [ ] No new warnings
- [x] New tests have been added
- [x] All tests pass locally with my changes

### CODE REVIEW

**Review by amandelpie — APPROVED** [review]

File: utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    override fun accept(type: FuzzedType) = !shouldPass(type.classId)
+
+    override fun generate(
+        description: FuzzedDescription,
+        type: FuzzedType
+    ) = sequence {
+        val classId = type.classId
+        // this adds mock support, but we should discuss first if mock is really needed
+//        if (description.description.shouldMock(classId)) {
```
Comment (denis-fokin) [inline utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt:34]: Let's remove comments

File: utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            },
+            empty = Routine.Empty {
+                UtNullModel(classId).fuzzed {
+                    summary = "%var% = null"
+                }
+            }
+        )
+    }
+
+    private fun shouldPass(type: ClassId): Boolean {
```
Comment (denis-fokin) [inline utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt:140]: What "should pass" means in this context?

File: utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+internal class FieldDescription(
+    val name: String,
+    val classId: ClassId,
+    val canBeSetDirectly: Boolean,
+    val setter: Method?,
+    val getter: Method?
+)
+
+internal fun findSuitableFields(classId: ClassId, packageName: String?): List<FieldDescription>  {
```
Comment (denis-fokin) [inline utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt:196]: I would add a piece of documentation here. Suitable for what?

File: utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt (hunk 2 above)
Comment (Markoutte) [reply utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt]: Changed to `isIgnored(type: ClassId)`

File: utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt (hunk 3 above)
Comment (Markoutte) [reply utbot-fuzzers/src/main/kotlin/org/utbot/fuzzing/providers/Objects.kt]: Changed to `findAccessibleModifableFields(...)`

File: utbot-fuzzing/src/main/kotlin/org/utbot/fuzzing/Api.kt [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+     * Fuzzing combines, randomize and mutates values using the seeds.
+     * Then it generates values and runs them with this method. This method should provide some feedback,
+     * which is the most important part for good fuzzing result. [emptyFeedback] can be provided only for test
+     * or infinite loops. Consider to implement own implementation of [Feedback] to provide more correct data or
+     * use [BaseFeedback] to generate key based feedback. In this case, the key is used to analyse what value should be next.
+     *
+     * @param description contains user-defined information about current run. Can be used as a state of the run.
+     * @param values current values to run.
+     */
+    suspend fun run(description: DESCRIPTION, values: List<RESULT>): FEEDBACK
```
Comment (denis-fokin) [inline utbot-fuzzing/src/main/kotlin/org/utbot/fuzzing/Api.kt:42]: Please, find a better naming for the method. 'Run' makes me think that the method is supposed to be called only ones. I would suggest something like: 'hanlde', 'valueUdated' or something similar

File: utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt [hunk 5]
```diff
@@ -387,89 +322,37 @@ class UtBotSymbolicEngine(
...
-        } else {
-            // in case a method with no parameters is passed fuzzing tries to fuzz this instance with different constructors, setters and field mutators
-            fuzz(syntheticMethodForFuzzingThisInstanceDescription, ObjectModelProvider(defaultIdGenerator).apply {
-                totalLimit = 500
-            })
-        }.withMutations(
-            TrieBasedFuzzerStatistics(coveredInstructionValues), methodUnderTestDescription, *defaultModelMutators().toTypedArray()
-        )
-        fuzzedValues.forEach { values ->
+        val names = graph.body.method.tags.filterIsInstance<ParamNamesTag>().firstOrNull()?.names ?: emptyList()
```
Comment (volivan239) [inline utbot-framework/src/main/kotlin/org/utbot/engine/UtBotSymbolicEngine.kt:340]: This `name` initialization looks weird. Can't we just use `methodUnderTest.parameters.map { it.name }` here? If not, please, add comment reasoning such an initialization

**Review by volivan239 — APPROVED** [review]
Looks monumental!

**Pull request merged**